packages/cli/src/modules/insights/insights.service.ts

```diff
@@ -89,68 +135,139 @@ export class InsightsService {
-				);
-			}
+		// time saved event
+		if (status === 'success' && ctx.workflowData.settings?.timeSavedPerExecution) {
+			this.bufferedInsights.add({
+				...commonWorkflowData,
+				type: 'time_saved_min',
+				value: ctx.workflowData.settings.timeSavedPerExecution,
+			});
+		}
 
-			await trx.upsert(
-				InsightsMetadata,
-				{
-					workflowId: ctx.workflowData.id,
-					workflowName: ctx.workflowData.name,
-					projectId: sharedWorkflow.projectId,
-					projectName: sharedWorkflow.project.name,
-				},
-				['workflowId'],
-			);
-			const metadata = await trx.findOneBy(InsightsMetadata, {
-				workflowId: ctx.workflowData.id,
+		if (!this.isAsynchronouslySavingInsights) {
+			// If we are not asynchronously saving insights, we need to flush the events
+			await this.flushEvents();
+		}
+
+		// If the buffer is full, flush the events asynchronously
+		if (this.bufferedInsights.size >= config.flushBatchSize) {
+			// Fire and forget flush to avoid blocking the workflow execute after handler
+			void this.flushEvents();
+		}
+	}
+
+	async saveInsightsMetadataAndRaw(insightsRawToInsertBuffer: Set<BufferedInsight>) {
+		const workflowIdNames: Map<string, string> = new Map();
+
+		for (const event of insightsRawToInsertBuffer) {
+			workflowIdNames.set(event.workflowId, event.workflowName);
+		}
+
+		await this.sharedWorkflowRepository.manager.transaction(async (trx) => {
+			const sharedWorkflows = await trx.find(SharedWorkflow, {
+				where: { workflowId: In([...workflowIdNames.keys()]), role: 'workflow:owner' },
+				relations: { project: true },
 			});
 
-			if (!metadata) {
-				// This can't happen, we just wrote the metadata in the same
-				// transaction.
-				throw new UnexpectedError(
-					`Could not find metadata for the workflow with the id '${ctx.workflowData.id}'`,
-				);
+			// Upsert metadata for the workflows that are not already in the cache or have
+			// different project or workflow names
+			const metadataToUpsert = sharedWorkflows.reduce((acc, workflow) => {
+				const cachedMetadata = this.cachedMetadata.get(workflow.workflowId);
+				if (
+					!cachedMetadata ||
+					cachedMetadata.projectId !== workflow.projectId ||
+					cachedMetadata.projectName !== workflow.project.name ||
+					cachedMetadata.workflowName !== workflowIdNames.get(workflow.workflowId)
+				) {
+					const metadata = new InsightsMetadata();
+					metadata.projectId = workflow.projectId;
+					metadata.projectName = workflow.project.name;
+					metadata.workflowId = workflow.workflowId;
+					metadata.workflowName = workflowIdNames.get(workflow.workflowId)!;
+
+					acc.push(metadata);
+				}
+				return acc;
+			}, [] as InsightsMetadata[]);
+
+			await trx.upsert(InsightsMetadata, metadataToUpsert, ['workflowId']);
+
+			const upsertMetadata = await trx.findBy(InsightsMetadata, {
+				workflowId: In(metadataToUpsert.map((m) => m.workflowId)),
```

Comment:
Optimization that may not be worth it, but I thought I'd point it out. We technically only need to update the ones that were not in the cache and outdated. But that may not be worth the complexity until we have prove it's an issue.

Reply:
@despairblue that's what `metadataToUpsert` is: metadata not in cached or outdated (different workflow name, project id or name), it's not the full workflow list (https://github.com/n8n-io/n8n/pull/14261/files#diff-f962282c430a8fc84dfad6b95a70c89f7cb4c342f5fc9b5bc1c6cd6c18c7a90bR195). 
Or I misunderstood your point :thinking:

Reply:
I misread the code. I thought it's doing `In([...workflowIdNames.keys()])`, but it's not 🤦🏾